suitcase/mongo_embedded/__init__.py

```diff
@@ -140,61 +151,123 @@ def __call__(self, name, doc):
+                        self._db_event_count['count_' + descriptor] += len(
+                                event_page['seq_num'])
+                last_push = time.monotonic()
+                do_push = False
+
+    @_try_wrapper
+    def _datum_worker(self):
+        # Gets datum from the queue, embedds them, and writes them to the
+        # volatile database.
+
+        last_push = 0
+        datum = None
 
+        # When a stop document is received 'False' is pushed on to the
+        # queue, this signals the worker to finish.
+        while datum is not False:
+            last_push = time.monotonic()
+            do_push = False
+            try:
+                if datum is None:
+                    datum = self._datum_queue.get(timeout=0.5)
+            except queue.Empty:
+                do_push = True
+            else:
+                # embedder.insert() returns None if the document is inserted,
+                # and returns the document, if embedder is full.
+                if datum is not False:
+                    datum = self._datum_embedder.insert(datum)
+            if (
+                    datum is not None
+                    or datum is False
+                    or do_push
+                    or time.monotonic() > (last_push + self._MAX_INSERT)):
+
+                if not self._datum_embedder.empty():
+                    datum_dump = self._datum_embedder.dump()
+                    self._bulkwrite_datum(datum_dump)
+                    for resource, datum_page in datum_dump.items():
+                        self._db_datum_count['count_' + resource] += len(
+                                datum_page['datum_id'])
+                last_push = time.monotonic()
+                do_push = False
+
+    @_try_wrapper
     def _count_worker(self):
         # Updates event_count and datum_count in the header document
 
         last_event_count = defaultdict(lambda: 0)
         last_datum_count = defaultdict(lambda: 0)
 
         while not self._frozen:
-            try:
-                time.sleep(5)
-
-                # Only updates the header if the count has changed.
-                if (
-                        (sum(self._db_event_count.values()) >
-                         sum(last_event_count.values()))
-                        or (sum(self._db_datum_count.values()) >
-                            sum(last_datum_count.values()))
-                   ):
-                    self._volatile_db.header.update_one(
-                        {'run_id': self._run_uid},
-                        {'$set': {**dict(self._db_event_count),
-                                  **dict(self._db_datum_count)}})
-
-                    last_event_count = self._db_event_count
-                    last_datum_count = self._db_datum_count
-            except BaseException as error:
-                self._datum_buffer.worker_error = error
-                self._event_buffer.worker_error = error
+            time.sleep(5)
```

Comment:
You might try a pattern like [this](https://github.com/NSLS-II/caproto/blob/b879e09ffb13031158a7ccb7caf8e6af0d82a4c4/caproto/threading/client.py#L577) to specifically signal "Time to clean up" as opposed to checking `self._frozen` here. It would enable to short-circuit the sleep when it's time to stop. You could use `event.wait(timeout=5)` instead of `time.sleep(5)` so that the loop moves on if 5 seconds passed _or if it's time to clean up_. Then you could flip `self._count_executor.shutdown(wait=False)` without imposing a long wait, and you would ensure that you don't leave any dangling threads.